agent/taskresource/grpcclient/credentialsfetcherclient_test.go

```diff
@@ -111,6 +131,105 @@ func TestCredentialsFetcherClient_AddKerberosLease(t *testing.T) {
 	}
 }
 
+func TestCredentialsFetcherClient_AddNonDomainJoinedKerberosLease(t *testing.T) {
+	tests := []struct {
+		name             string
+		credspecContents []string
+		username         string
+		password         string
+		domain           string
+		response         CredentialsFetcherResponse
+	}{
+		{
+			"invalid request empty credspec contents",
+			[]string{},
+			"testusername",
+			"testpassword",
+			"testdomain",
+			CredentialsFetcherResponse{},
+		},
+		{
+			"invalid request username, password or domain should not be empty",
+			[]string{credspec_webapp01},
+			"",
+			"",
+			"",
+			CredentialsFetcherResponse{},
+		},
+		{
+			"valid request credspecs associated to gMSA account",
+			[]string{credspec_webapp01},
+			"testusername",
+			"testpassword",
+			"testdomain",
+			CredentialsFetcherResponse{LeaseID: leaseid, KerberosTicketPaths: []string{"/var/credentials-fetcher/krbdir/123456/webapp01", "/var/credentials-fetcher/krbdir/123456/webapp02"}},
+		},
+	}
+
+	ctx := context.Background()
+
+	conn, err := grpc.DialContext(ctx, "", grpc.WithInsecure(), grpc.WithContextDialer(dialer()))
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer conn.Close()
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			response, _ := NewCredentialsFetcherClient(conn, time.Minute).AddNonDomainJoinedKerberosLease(context.Background(), tt.credspecContents, tt.username, tt.password, tt.domain)
+			if response.LeaseID != tt.response.LeaseID {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
```

Comment:
We should not branch the assertions based on the output of the function being tested which is `response` in this case. The assertions should be based on the test case alone.

How about adding an `expectedError string` field to the test cases and use it for branching the assertions? This way we can also assert on the errors returned by the function being tested. This is a follow-up to my earlier comment about making `AddNonDomainJoinedKerberosLease` method return an error is validation fails.

```go
if tt.expectedError != "" {
    assert.EqualError(t, err, tt.expectedError)
} else {
    assert.Equal(t, tt.response, response)
}
```

Same comment for `TestCredentialsFetcherClient_RenewNonDomainJoinedKerberosLease` below.

Reply:
will add the conditions accordingly